datadog/dogstatsd/container.py

```diff
@@ -13,31 +14,51 @@ class UnresolvableContainerID(Exception):
     """
 
 
-class ContainerID(object):
+class Cgroup(object):
     """
-    A reader class that retrieves the current container ID parsed from a the cgroup file.
+    A reader class that retrieves either:
+    - The current container ID parsed from the cgroup file
+    - The cgroup controller inode.
 
     Returns:
-    object: ContainerID
+    object: Cgroup
 
     Raises:
         `NotImplementedError`: No proc filesystem is found (non-Linux systems)
         `UnresolvableContainerID`: Unable to read the container ID
     """
 
     CGROUP_PATH = "/proc/self/cgroup"
+    CGROUP_MOUNT_PATH = "/sys/fs/cgroup"  # cgroup mount path.
+    CGROUP_NS_PATH = "/proc/self/ns/cgroup"  # path to the cgroup namespace file.
+    CGROUPV1_BASE_CONTROLLER = "memory"  # controller used to identify the container-id in cgroup v1 (memory).
+    CGROUPV2_BASE_CONTROLLER = ""  # controller used to identify the container-id in cgroup v2.
+    HOST_CGROUP_NAMESPACE_INODE = "0xEFFFFFFBL"  # inode of the host cgroup namespace.
+
     UUID_SOURCE = r"[0-9a-f]{8}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{12}"
     CONTAINER_SOURCE = r"[0-9a-f]{64}"
     TASK_SOURCE = r"[0-9a-f]{32}-\d+"
     LINE_RE = re.compile(r"^(\d+):([^:]*):(.+)$")
     CONTAINER_RE = re.compile(r"(?:.+)?({0}|{1}|{2})(?:\.scope)?$".format(UUID_SOURCE, CONTAINER_SOURCE, TASK_SOURCE))
 
     def __init__(self):
-        self.container_id = self._read_container_id(self.CGROUP_PATH)
+        if self._is_cgroup_namespace():
+            self.container_id = self._read_cgroup_path()
+            return
+        self.container_id = self._get_cgroup_from_inode()
+
+    def _is_cgroup_namespace(self):
+        """Check if the current process is in a host cgroup namespace."""
+        return (
+            os.stat(self.CGROUP_NS_PATH).st_ino == self.HOST_CGROUP_NAMESPACE_INODE
```

Comment:
Could there be a case where this raises an exception and we would want to pursue with getting the cgroup controller inode ?

Reply:
I don't think we have any specific exception to catch here to resolve it. If it does fail, it will be catched here: https://github.com/DataDog/datadogpy/blob/master/datadog/dogstatsd/base.py#L1290C17-L1290C17

Maybe we can get a confirmation about this from @vickenty?

Reply:
We probably want the logic to be exactly the same between all clients, typically Go is the "reference" implementation. It seems Go treats errors in `isHostCgroupNamespace` as "false", and skips to inode code. Let's do the same here.

Reply:
I've added a `try/except`
```diff
diff --git a/datadog/dogstatsd/container.py b/datadog/dogstatsd/container.py
index fc911cf..94b5be2 100644
--- a/datadog/dogstatsd/container.py
+++ b/datadog/dogstatsd/container.py
@@ -49,11 +49,14 @@ class Cgroup(object):

     def _is_host_cgroup_namespace(self):
         """Check if the current process is in a host cgroup namespace."""
-        return (
-            os.stat(self.CGROUP_NS_PATH).st_ino == self.HOST_CGROUP_NAMESPACE_INODE
-            if os.path.exists(self.CGROUP_NS_PATH)
-            else False
-        )
+        try:
+            return (
+                os.stat(self.CGROUP_NS_PATH).st_ino == self.HOST_CGROUP_NAMESPACE_INODE
+                if os.path.exists(self.CGROUP_NS_PATH)
+                else False
+            )
+        except Exception:
+            return False

     def _read_cgroup_path(self):
         """Read the container ID from the cgroup file."""
```
There is no logging here though.